packager/VisibilityChecker.cc

```diff
@@ -789,39 +843,52 @@ class VisibilityCheckerPass final {
                 ast::TreeWalk::apply(ctx, pass, f.tree);
-                resultq->push(make_pair(f.file, std::move(pass.packageReferences)), 1);
+                resultq->push(make_tuple(f.file, std::move(pass.packageReferences), std::move(pass.symbolsReferenced)),
+                              1);
             }
             barrier.DecrementCount();
         });
 
         if (gs.packageDB().genPackages()) {
-            for (auto &parsedFile : filesSpan) {
-                // TODO(neil): we can skip this loop on the slow path since the package references sets will be empty
-                auto file = parsedFile.file;
-                auto pkgName = gs.packageDB().getPackageNameForFile(file);
-                if (!pkgName.exists()) {
-                    continue;
-                }
-                auto nonConstPackageInfo = nonConstPackageDB.getPackageInfoNonConst(pkgName);
-                nonConstPackageInfo->untrackPackageReferencesFor(file);
+            {
+                Timer timeit(gs.tracer(), "visibility_checker.untrackPackageReferences");
+                untrackPackageReferences(nonConstPackageDB, filesSpan);
+            }
+
+            {
+                Timer timeit(gs.tracer(), "visibility_checker.untrackSymbolReferences");
+                untrackSymbolReferences(nonConstGs, filesSpan);
             }
 
-            std::optional<std::pair<core::FileRef,
-                                    UnorderedMap<core::packages::MangledName, core::packages::PackageReferenceInfo>>>
+            std::optional<std::tuple<core::FileRef,
+                                     UnorderedMap<core::packages::MangledName, core::packages::PackageReferenceInfo>,
+                                     UnorderedSet<core::SymbolRef>>>
                 threadResult;
             for (auto result = resultq->wait_pop_timed(threadResult, WorkerPool::BLOCK_INTERVAL(), gs.tracer());
                  !result.done();
                  result = resultq->wait_pop_timed(threadResult, WorkerPool::BLOCK_INTERVAL(), gs.tracer())) {
                 if (result.gotItem() && threadResult.has_value()) {
-                    auto file = threadResult.value().first;
+                    auto file = std::get<0>(threadResult.value());
                     auto pkgName = gs.packageDB().getPackageNameForFile(file);
                     if (!pkgName.exists()) {
                         continue;
                     }
                     auto nonConstPackageInfo = nonConstPackageDB.getPackageInfoNonConst(pkgName);
-                    for (auto [p, packageReferenceInfo] : threadResult.value().second) {
+                    for (auto [p, packageReferenceInfo] : std::get<1>(threadResult.value())) {
                         nonConstPackageInfo->trackPackageReference(file, p, packageReferenceInfo);
                     }
+
+                    for (auto symbol : std::get<2>(threadResult.value())) {
```

Comment:
As a reader, the `get<2>` here makes this hard to understand, would you mind doing something that allows names, either a struct with named fields, or a structured binding, or assign the result of `get<2>` into a variable first, etc.?